File: src/chrome/browser/android/compositor/scene_layer/ephemeral_tab_scene_layer.cc
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/compositor/scene_layer/ephemeral_tab_scene_layer.h"
#include "base/android/jni_android.h"
#include "base/android/jni_array.h"
#include "cc/layers/solid_color_layer.h"
#include "chrome/browser/android/compositor/layer/ephemeral_tab_layer.h"
#include "content/public/browser/android/compositor.h"
#include "content/public/browser/web_contents.h"
#include "jni/EphemeralTabSceneLayer_jni.h"
#include "ui/android/resources/resource_manager_impl.h"
#include "ui/android/view_android.h"
#include "ui/gfx/android/java_bitmap.h"

using base::android::JavaParamRef;
using base::android::JavaRef;

namespace android {

EphemeralTabSceneLayer::EphemeralTabSceneLayer(JNIEnv* env,
                                               const JavaRef<jobject>& jobj)
    : SceneLayer(env, jobj),
      base_page_brightness_(1.0f),
      content_container_(cc::Layer::Create()) {
  // Responsible for moving the base page without modifying the layer itself.
  content_container_->SetIsDrawable(true);
  content_container_->SetPosition(gfx::PointF(0.0f, 0.0f));
  layer()->AddChild(content_container_);
}

EphemeralTabSceneLayer::~EphemeralTabSceneLayer() {}

void EphemeralTabSceneLayer::CreateEphemeralTabLayer(
    JNIEnv* env,
    const JavaParamRef<jobject>& object,
    const JavaParamRef<jobject>& jresource_manager) {
  ui::ResourceManager* resource_manager =
      ui::ResourceManagerImpl::FromJavaObject(jresource_manager);
  ephemeral_tab_layer_ = EphemeralTabLayer::Create(resource_manager);

  // The layer is initially invisible.
  ephemeral_tab_layer_->layer()->SetHideLayerAndSubtree(true);
  layer()->AddChild(ephemeral_tab_layer_->layer());
}

void EphemeralTabSceneLayer::SetResourceIds(JNIEnv* env,
                                            const JavaParamRef<jobject>& object,
                                            jint text_resource_id,
                                            jint bar_background_resource_id,
                                            jint bar_shadow_resource_id,
                                            jint panel_icon_resource_id,
                                            jint close_icon_resource_id) {
  ephemeral_tab_layer_->SetResourceIds(
      text_resource_id, bar_background_resource_id, bar_shadow_resource_id,
      panel_icon_resource_id, close_icon_resource_id);
}

void EphemeralTabSceneLayer::Update(JNIEnv* env,
                                    const JavaParamRef<jobject>& object,
                                    jint title_view_resource_id,
                                    jint caption_view_resource_id,
                                    jfloat caption_animation_percentage,
                                    jfloat text_layer_min_height,
                                    jfloat title_caption_spacing,
                                    jboolean caption_visible,
                                    jint progress_bar_background_resource_id,
                                    jint progress_bar_resource_id,
                                    jfloat dp_to_px,
                                    jfloat base_page_brightness,
                                    jfloat base_page_offset,
                                    const JavaParamRef<jobject>& jweb_contents,
                                    jfloat panel_x,
                                    jfloat panel_y,
                                    jfloat panel_width,
                                    jfloat panel_height,
                                    jint bar_background_color,
                                    jfloat bar_margin_side,
                                    jfloat bar_height,
                                    jboolean bar_border_visible,
                                    jfloat bar_border_height,
                                    jboolean bar_shadow_visible,
                                    jfloat bar_shadow_opacity,
                                    jint icon_color,
                                    jboolean progress_bar_visible,
                                    jfloat progress_bar_height,
                                    jfloat progress_bar_opacity,
                                    jint progress_bar_completion) {
  // NOTE(mdjones): It is possible to render the panel before content has been
  // created. If this is the case, do not attempt to access the WebContents
  // and instead pass null.
  content::WebContents* web_contents =
      content::WebContents::FromJavaWebContents(jweb_contents);
  scoped_refptr<cc::Layer> content_layer =
      web_contents ? web_contents->GetNativeView()->GetLayer() : nullptr;
  // Fade the base page out.
  if (base_page_brightness_ != base_page_brightness) {
    base_page_brightness_ = base_page_brightness;
    cc::FilterOperations filters;
    if (base_page_brightness < 1.f) {
      filters.Append(
          cc::FilterOperation::CreateBrightnessFilter(base_page_brightness));
    }
    content_container_->SetFilters(filters);
  }
  // Move the base page contents up.
  content_container_->SetPosition(gfx::PointF(0.0f, base_page_offset));
  ephemeral_tab_layer_->SetProperties(
      title_view_resource_id, caption_view_resource_id,
      caption_animation_percentage, text_layer_min_height,
      title_caption_spacing, caption_visible,
      progress_bar_background_resource_id, progress_bar_resource_id, dp_to_px,
      content_layer, panel_x, panel_y, panel_width, panel_height,
      bar_background_color, bar_margin_side, bar_height, bar_border_visible,
      bar_border_height, bar_shadow_visible, bar_shadow_opacity, icon_color,
      progress_bar_visible, progress_bar_height, progress_bar_opacity,
      progress_bar_completion);
  // Make the layer visible if it is not already.
  ephemeral_tab_layer_->layer()->SetHideLayerAndSubtree(false);
}

void EphemeralTabSceneLayer::SetContentTree(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj,
    const JavaParamRef<jobject>& jcontent_tree) {
  SceneLayer* content_tree = FromJavaObject(env, jcontent_tree);
  if (!content_tree || !content_tree->layer())
    return;
  if (!content_tree->layer()->parent() ||
      (content_tree->layer()->parent()->id() != content_container_->id())) {
    content_container_->AddChild(content_tree->layer());
  }
}

void EphemeralTabSceneLayer::HideTree(JNIEnv* env,
                                      const JavaParamRef<jobject>& jobj) {
  // TODO(mdjones): Create super class for this logic.
  if (ephemeral_tab_layer_) {
    ephemeral_tab_layer_->layer()->SetHideLayerAndSubtree(true);
  }
  // Reset base page brightness.
  cc::FilterOperations filters;
  filters.Append(cc::FilterOperation::CreateBrightnessFilter(1.0f));
  content_container_->SetFilters(filters);
  // Reset base page offset.
  content_container_->SetPosition(gfx::PointF(0.0f, 0.0f));
}

static jlong JNI_EphemeralTabSceneLayer_Init(
    JNIEnv* env,
    const JavaParamRef<jobject>& jobj) {
  // This will automatically bind to the Java object and pass ownership there.
  EphemeralTabSceneLayer* ephemeral_tab_scene_layer =
      new EphemeralTabSceneLayer(env, jobj);
  return reinterpret_cast<intptr_t>(ephemeral_tab_scene_layer);
}

}  // namespace android

File: chrome/android/java/src/org/chromium/chrome/browser/compositor/scene_layer/EphemeralTabSceneLayer.java
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.compositor.scene_layer;

import android.support.annotation.Nullable;
import android.text.TextUtils;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.chrome.R;
import org.chromium.chrome.browser.ChromeFeatureList;
import org.chromium.chrome.browser.compositor.bottombar.ephemeraltab.EphemeralTabBarControl;
import org.chromium.chrome.browser.compositor.bottombar.ephemeraltab.EphemeralTabCaptionControl;
import org.chromium.chrome.browser.compositor.bottombar.ephemeraltab.EphemeralTabPanel;
import org.chromium.chrome.browser.compositor.bottombar.ephemeraltab.EphemeralTabTitleControl;
import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.content_public.browser.WebContents;
import org.chromium.ui.resources.ResourceManager;

/**
 * A SceneLayer to render layers for Ephemeral Tab.
 */
@JNINamespace("android")
public class EphemeralTabSceneLayer extends SceneOverlayLayer {
    /** Pointer to native EphemeralTabSceneLayer. */
    private long mNativePtr;

    /** If the scene layer has been initialized. */
    private boolean mIsInitialized;

    /** The conversion multiple from dp to px. */
    private final float mDpToPx;

    private final int mFaviconSizePx;

    private String mCachedUrl;

    /**
     * @param dpToPx The conversion multiple from dp to px for the device.
     * @param faviconSizeDp Preferred size of the favicon to fetch.
     */
    public EphemeralTabSceneLayer(float dpToPx, int faviconSizeDp) {
        mDpToPx = dpToPx;
        mFaviconSizePx = (int) (faviconSizeDp * dpToPx);
    }

    /**
     * Update the scene layer to draw an OverlayPanel.
     * @param resourceManager Manager to get view and image resources.
     * @param panel The OverlayPanel to render.
     * @param bar {@link EphemeralTabBarControl} object.
     * @param title {@link EphemeralTabTitleControl} object.
     * @param caption {@link EphemeralTabCaptionControl} object.
     */
    public void update(ResourceManager resourceManager, EphemeralTabPanel panel,
            EphemeralTabBarControl bar, EphemeralTabTitleControl title,
            @Nullable EphemeralTabCaptionControl caption) {
        // Don't try to update the layer if not initialized or showing.
        if (resourceManager == null || !panel.isShowing()) return;
        if (!mIsInitialized) {
            nativeCreateEphemeralTabLayer(mNativePtr, resourceManager);
            int openInTabIconId = (ChromeFeatureList.isEnabled(ChromeFeatureList.OVERLAY_NEW_LAYOUT)
                                          && panel.canPromoteToNewTab())
                    ? R.drawable.open_in_new_tab
                    : INVALID_RESOURCE_ID;
            int dragHandlebarId = ChromeFeatureList.isEnabled(ChromeFeatureList.OVERLAY_NEW_LAYOUT)
                    ? R.drawable.drag_handlebar
                    : INVALID_RESOURCE_ID;
            nativeSetResourceIds(mNativePtr, title.getViewId(),
                    R.drawable.contextual_search_bar_background, R.drawable.modern_toolbar_shadow,
                    R.drawable.infobar_chrome, dragHandlebarId, openInTabIconId,
                    R.drawable.btn_close);
            mIsInitialized = true;
        }

        int titleViewId = title.getViewId();
        int captionViewId = 0;
        float captionAnimationPercentage = 0.f;
        boolean captionVisible = false;
        if (caption != null) {
            captionViewId = caption.getViewId();
            captionAnimationPercentage = caption.getAnimationPercentage();
            captionVisible = caption.getIsVisible();
        }
        boolean isProgressBarVisible = panel.isProgressBarVisible();
        float progressBarHeight = panel.getProgressBarHeight();
        float progressBarOpacity = panel.getProgressBarOpacity();
        int progressBarCompletion = panel.getProgressBarCompletion();

        WebContents panelWebContents = panel.getWebContents();
        nativeUpdate(mNativePtr, titleViewId, captionViewId, captionAnimationPercentage,
                bar.getTextLayerMinHeight(), bar.getTitleCaptionSpacing(), captionVisible,
                R.drawable.progress_bar_background, R.drawable.progress_bar_foreground, mDpToPx,
                panel.getBasePageBrightness(), panel.getBasePageY() * mDpToPx, panelWebContents,
                panel.getOffsetX() * mDpToPx, panel.getOffsetY() * mDpToPx,
                panel.getWidth() * mDpToPx, panel.getHeight() * mDpToPx,
                panel.getBarBackgroundColor(), panel.getBarMarginSide() * mDpToPx,
                panel.getBarMarginTop() * mDpToPx, panel.getBarHeight() * mDpToPx,
                panel.isBarBorderVisible(), panel.getBarBorderHeight() * mDpToPx,
                panel.getBarShadowVisible(), panel.getBarShadowOpacity(), panel.getIconColor(),
                panel.getDragHandlebarColor(), isProgressBarVisible, progressBarHeight * mDpToPx,
                progressBarOpacity, progressBarCompletion);

        String url = panel.getUrl();
        if (!TextUtils.equals(mCachedUrl, url)) {
            nativeGetFavicon(mNativePtr, Profile.getLastUsedProfile(), url, mFaviconSizePx);
            mCachedUrl = url;
        }
    }

    @Override
    public void setContentTree(SceneLayer contentTree) {
        nativeSetContentTree(mNativePtr, contentTree);
    }

    /**
     * Hide the layer tree; for use if the panel is not being shown.
     */
    public void hideTree() {
        if (!mIsInitialized) return;
        nativeHideTree(mNativePtr);
        mCachedUrl = null;
    }

    @Override
    protected void initializeNative() {
        if (mNativePtr == 0) {
            mNativePtr = nativeInit();
        }
        assert mNativePtr != 0;
    }

    /**
     * Destroys this object and the corresponding native component.
     */
    @Override
    public void destroy() {
        super.destroy();
        mIsInitialized = false;
        mNativePtr = 0;
    }

    private native long nativeInit();
    private native void nativeCreateEphemeralTabLayer(
            long nativeEphemeralTabSceneLayer, ResourceManager resourceManager);
    private native void nativeSetContentTree(
            long nativeEphemeralTabSceneLayer, SceneLayer contentTree);
    private native void nativeHideTree(long nativeEphemeralTabSceneLayer);
    private native void nativeSetResourceIds(long nativeEphemeralTabSceneLayer,
            int barTextResourceId, int barBackgroundResourceId, int barShadowResourceId,
            int panelIconResourceId, int dragHandlebarResourceId, int openTabIconResourceId,
            int closeIconResourceId);
    private native void nativeGetFavicon(
            long nativeEphemeralTabSceneLayer, Profile profile, String url, int size);
    private native void nativeUpdate(long nativeEphemeralTabSceneLayer, int titleViewId,
            int captionViewId, float captionAnimationPercentage, float textLayerMinHeight,
            float titleCaptionSpacing, boolean captionVisible, int progressBarBackgroundResourceId,
            int progressBarResourceId, float dpToPx, float basePageBrightness,
            float basePageYOffset, WebContents webContents, float panelX, float panelY,
            float panelWidth, float panelHeight, int barBackgroundColor, float barMarginSide,
            float barMarginTop, float barHeight, boolean barBorderVisible, float barBorderHeight,
            boolean barShadowVisible, float barShadowOpacity, int iconColor, int dragHandlebarColor,
            boolean isProgressBarVisible, float progressBarHeight, float progressBarOpacity,
            int progressBarCompletion);
}
